lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.h

```diff
@@ -0,0 +1,83 @@
+
+#ifndef liblldb_TypeRefCacher_h_
+#define liblldb_TypeRefCacher_h_
+
+#include <unordered_map>
+
+#include "lldb/Core/Module.h"
+
+#include "llvm/ADT/STLExtras.h"
+
+#include "swift/Reflection/ReflectionContext.h"
+
+
+namespace lldb_private {
+struct SwiftMetadataCache {
+public:
+  static SwiftMetadataCache &get(); 
+  SwiftMetadataCache(SwiftMetadataCache &) = delete;
+  void operator=(SwiftMetadataCache const&) = delete;
+
+  /// Relate the lldb module with the reflection info id.
+  void registerModuleWithReflectionInfoID(lldb::ModuleSP module,
+                                          uint64_t info_id);
+
+  /// Cache the field descriptors in the info with the mangled names
+  /// passed in. The number of field descriptors and mangled names passed should
+  /// be the same, otherwise caching is aborted.
+  void cacheFieldDescriptors(
+      uint64_t info_id,
+      const swift::reflection::FieldSection &field_descriptors,
+      const std::vector<std::string> &mangled_names);
+
+  llvm::Optional<std::pair<uint64_t, uint64_t>> 
+    getFieldDescriptor(const std::string &mangled_name) const;
+
+  /// Returns whether looking for a mangled name in reflection info with a given
+  /// id could possibly return information that wasn't already in the cache.
+  bool shouldParseReflectionInfoWithId(uint64_t info_id) const;
+
+private:
+  SwiftMetadataCache() = default;
+
+  void loadCacheForModule(lldb::ModuleSP module, uint64_t info_id);
+
+  bool writeMangledNamesAndOffsetsToEncoder(
+      uint64_t info_id,
+      const swift::reflection::FieldSection &field_descriptors,
+      const std::vector<std::string> &mangled_names, DataEncoder &encoder);
+
+  void fillMangledNameToOffsetMap(
+    uint64_t info_id, const swift::reflection::ReflectionInfo &info,
+    const std::vector<std::string> &mangled_names);
+
+  /// Gets the file name that the cache file should use for a given module.
+  std::string
+  getTyperefCacheFileNameForModule(const lldb::ModuleSP &module);
+
+  /// A map from mangled names to a pair constituting of the reflection info's
+  /// id and the offset of the field descriptor in the field section buffer it's
+  /// in.
+  llvm::StringMap<std::pair<uint64_t, uint64_t>> m_mangled_name_to_offset;
+
+  /// A map from reflection infos ids to a pair constituting of its
+  /// corresponding module and whether or not we've inserted the cached metadata
+  /// for that reflection info already.
+  std::unordered_map<uint64_t, std::pair<lldb::ModuleSP, bool>>
+      m_info_to_module;
+
+  /// Version number that indicates how the data is laid out.
+  /// Version 1 is encoded as follows:
+  /// A header constituting of:
+  /// - Version number (uint16_t).
+  /// - Module signature.
```

Comment:
What exactly is the module signature? And how long is it?

Reply:
The module signature is its UUID, which varies depending on the object file format. For Mach-O that's what's specified in the `LC_UUID`, and is 16 bytes.